File: Bliki-OldBlog/DebuggerDisplay/Programming2012-03-28/DebuggerDisplay/DebuggerDisplay/Person.vb
Imports System.ComponentModel

<DebuggerDisplay("{DebuggerDisplay}")> _
Public Class SuperHero
    Inherits Person

    Private _power As String
    Public ReadOnly Property Power() As String
        Get
            Return _power
        End Get
    End Property

    Public Sub New(ByVal name As String, ByVal age As Integer, ByVal power As String)
        MyBase.New(name, age)
        _power = power
    End Sub

    Private ReadOnly Property DebuggerDisplay() As String
        Get
            Return "{Name} is not just a human!"
        End Get
    End Property
End Class

<DebuggerDisplay("{DebuggerDisplay")> _
Public Class Person
    Private _name As String
    Private _age As Integer
    Private _contact As ContactDetails

    Public ReadOnly Property Name() As String
        Get
            Return _name
        End Get
    End Property

    Public ReadOnly Property Age() As Integer
        Get
            Return _age
        End Get
    End Property

    Public ReadOnly Property Contact() As ContactDetails
        Get
            Return _contact
        End Get
    End Property

    Public Sub New(ByVal name As String, ByVal age As Integer)
        Me.New(name, age, Nothing)
    End Sub

    Public Sub New(ByVal name As String, ByVal age As Integer, ByVal email As String, ByVal tel As String)
        Me.New(name, age, New ContactDetails(email, tel))
    End Sub

    Public Sub New(ByVal name As String, ByVal age As Integer, ByVal contact As ContactDetails)
        _name = name
        _age = age
        _contact = contact
    End Sub

    Private ReadOnly Property DebuggerDisplay() As String
        Get
            Return "This person is {Name}. {If(Age < 20, ""Young"", Age)}"
        End Get
    End Property
End Class

<DebuggerDisplay("{ToString(),nq} =")> _
Public Enum ContactTypes
    <Description("Unknown address")> _
    Unknown
    <Description("Home address")> _
    Home
    <Description("Work address")> _
    Work
    <Description("Billing address")> _
    Billing
End Enum

Public Class ContactDetails
    Private _email As String
    Private _telephone As String

    Public ReadOnly Property Email() As String
        Get
            Return _email
        End Get
    End Property

    Public ReadOnly Property Telephone() As String
        Get
            Return _telephone
        End Get
    End Property

    Public Sub New(ByVal email As String, ByVal tel As String)
        _email = email
        _telephone = tel
    End Sub
End Class

File: Bliki-OldBlog/DebuggerDisplay/Programming2012-03-28/DebuggerDisplayCSharp/DebuggerDisplayCSharp/Person.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;
using System.ComponentModel;
using System.Reflection;

namespace DebuggerDisplayCSharp
{
	//[DebuggerDisplay(@"Capacity=\{{Length,nq} / {Capacity,nq}\}")]
	[DebuggerDisplay("MemoryUsage=\\{{Length,nq} / {Capacity,nq}\\} ({(Length / (double)Capacity).ToString(\"p\"),nq})")]
	public class Test
	{
		public int Length { get; set; }
		public int Capacity { get; set; }

		public Test()
		{
			Length = 5;
			Capacity = 10;
		}

		public override string ToString()
		{
			return base.ToString();
		}
	}

	[DebuggerDisplay(@"{DebuggerDisplay}")]
	public class SuperHero : Person
	{
		public string Power { get; set; }

		public SuperHero(string name, int age, string power)
			: base(name, age)
		{
			Power = power;
		}

		private string DebuggerDisplay
		{
			get
			{
				return string.Format("{0} ({1})", base.DebuggerDisplay, Power);
			}
		}
	}


	//[Description("Test123"), DebuggerDisplay("This person is {Name}. {Age < 50 ? \"Young\" : \"Old\",nq}")]
	[DebuggerDisplay(@"{DebuggerDisplay}"), DebuggerTypeProxy(typeof(Person.PersonProxy))]
	public class Person
	{
		[EditorBrowsable(EditorBrowsableState.Advanced)]
		protected internal class PersonProxy
		{
			//[DebuggerBrowsable(DebuggerBrowsableState.Never)]
			private Person _p;

			//public Person Person { get { return _p; } }

			public PersonProxy(Person p)
			{
				_p = p;
			}

			
		}


		public virtual string Name { get; set; }
		public int Age { get; set; }
		public ContactDetails Contact { get; set; }

		public Person(string name, int age)
		{
			Name = name;
			Age = age;
		}

		public override string ToString()
		{
			return string.Format("ToString: {0}", Name);
		}

		[DebuggerBrowsable(DebuggerBrowsableState.Never)]
		protected virtual string DebuggerDisplay
		{
			get { return "DebuggerDisplay: " + Name + " (" + Age.ToString() + ")"; }
		}


	}

	[DebuggerDisplay("{DebuggerDisplay,nq}")]
	public sealed class Student
	{
		public string FirstName { get; set; }
		public string LastName { get; set; }
		private string DebuggerDisplay
		{
			get { return string.Format("Student: {0} {1}", FirstName, LastName); }
		}
	}


	public class ContactDetails
	{
		public string Email { get; set; }
		public string Telephone { get; set; }
		public string Address { get; set; }

		public ContactDetails(string email, string tel, string address)
		{
			Email = email;
			Telephone = tel;
			Address = address;
		}
	}

	public static class Helpers
	{
		public static string DebuggerDisplay(object value)
		{
			FieldInfo field = typeof(ContactTypes).GetField(value.ToString());
         return field.GetCustomAttributes(typeof(DescriptionAttribute), false)
                     .Cast<DescriptionAttribute>()
                     .Select(x => x.Description)
                     .FirstOrDefault();
		}
	}

	[DebuggerDisplay("{DebuggerDisplayCSharp.Helpers.DebuggerDisplay(this)}")]
	public enum ContactTypes
	{
		[Description("Uknown contacttype")]
		Unknown = 0,
		[Description("Home address")]
		Home = 1,
		[Description("Workplace address")]
		Work = 2,
		[Description("Parental address")]
		Billing = 4
	}
}
